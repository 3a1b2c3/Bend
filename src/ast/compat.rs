// TODO: Refactor to not use this intermediate form

use hvm_core::{Val, OP};

#[derive(Clone, Debug)]
/// Net representation used only as an intermediate for converting to hvm-core format
pub struct INet {
  pub nodes: Vec<Val>,
}

pub type NodeKind = Val;
pub type Port = Val;
pub type NodeId = Val;
pub type SlotId = Val;

/// The ROOT port is on the deadlocked root node at address 0.
pub const ROOT: Port = 1;
pub const TAG_WIDTH: u32 = 16; // TODO: Make this generic over the HVM type.
pub const TAG: u32 = Val::BITS - TAG_WIDTH;
pub const ERA: NodeKind = 0 << TAG;
pub const CON: NodeKind = 1 << TAG;
pub const DUP: NodeKind = 2 << TAG;
pub const REF: NodeKind = 3 << TAG;
pub const NUM_U32: NodeKind = 4 << TAG;
pub const NUM_I32: NodeKind = 5 << TAG;
pub const NUMOP: NodeKind = 6 << TAG;
pub const LABEL_MASK: NodeKind = (1 << TAG) - 1;
pub const TAG_MASK: NodeKind = !LABEL_MASK;

/// Create a new net, with a deadlocked root node.
pub fn new_inet() -> INet {
  INet {
    nodes: vec![2, 1, 0, ERA], // p2 points to p0, p1 points to net
  }
}

/// Allocates a new node, reclaiming a freed space if possible.
pub fn new_node(inet: &mut INet, kind: NodeKind) -> NodeId {
  let node = addr(inet.nodes.len() as Port);
  inet.nodes.extend([port(node, 0), port(node, 1), port(node, 2), kind]);
  node
}

/// Builds a port (an address / slot pair).
pub fn port(node: NodeId, slot: SlotId) -> Port {
  (node << 2) | slot
}

/// Returns the address of a port (TODO: rename).
pub fn addr(port: Port) -> NodeId {
  port >> 2
}

/// Returns the slot of a port.
pub fn slot(port: Port) -> SlotId {
  port & 3
}

/// Enters a port, returning the port on the other side.
pub fn enter(inet: &INet, port: Port) -> Port {
  inet.nodes[port as usize]
}

/// Kind of the node.
pub fn kind(inet: &INet, node: NodeId) -> NodeKind {
  inet.nodes[port(node, 3) as usize]
}

/// Links two ports.
pub fn link(inet: &mut INet, ptr_a: Port, ptr_b: Port) {
  inet.nodes[ptr_a as usize] = ptr_b;
  inet.nodes[ptr_b as usize] = ptr_a;
}

pub fn op_to_label(value: OP) -> NodeKind {
  match value {
    OP::ADD => 0x0,
    OP::SUB => 0x1,
    OP::MUL => 0x2,
    OP::DIV => 0x3,
    OP::MOD => 0x4,
    OP::EQ => 0x5,
    OP::NEQ => 0x6,
    OP::LT => 0x7,
    OP::GT => 0x8,
    OP::LTE => 0x9,
    OP::GTE => 0xa,
    OP::AND => 0xb,
    OP::OR => 0xc,
  }
}

pub fn label_to_op(value: NodeKind) -> OP {
  match value {
    0x0 => OP::ADD,
    0x1 => OP::SUB,
    0x2 => OP::MUL,
    0x3 => OP::DIV,
    0x4 => OP::MOD,
    0x5 => OP::EQ,
    0x6 => OP::NEQ,
    0x7 => OP::LT,
    0x8 => OP::GT,
    0x9 => OP::LTE,
    0xa => OP::GTE,
    0xb => OP::AND,
    0xc => OP::OR,
    _ => unreachable!(),
  }
}
#[derive(Debug)]
pub struct INode {
  pub kind: NodeKind,
  pub ports: [String; 3],
}

pub type INodes = Vec<INode>;
