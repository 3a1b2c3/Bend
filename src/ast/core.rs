use std::{collections::HashMap, fmt};

use super::DefId;
pub struct Book {
  pub defs: HashMap<DefId, LNet>,
}

// TODO: Import this module directly from hvm2

#[derive(Debug)]
pub struct LNet {
  root: LTree,
  acts: LActs,
}

#[derive(Debug)]
pub enum LTree {
  Era,
  Nod { tag: u8, lft: Box<LTree>, rgt: Box<LTree> },
  Var { nam: String },
  Num { val: u32 },
  Ref { nam: u32 },
}

type LActs = Vec<(LTree, LTree)>;

pub fn name_to_letters(name: &str) -> Vec<u8> {
  let mut letters = Vec::new();
  for c in name.chars() {
    letters.push(match c {
      '.' => 0,
      '0' ..= '9' => c as u8 - b'0' + 1,
      'A' ..= 'Z' => c as u8 - b'A' + 11,
      'a' ..= 'z' => c as u8 - b'a' + 37,
      '_' => 63,
      _ => panic!("Invalid character in name"),
    });
  }
  letters
}

pub fn letters_to_name(letters: Vec<u8>) -> String {
  let mut name = String::new();
  for letter in letters {
    name.push(match letter {
      0 => '.',
      1 ..= 10 => (letter - 1 + b'0') as char,
      11 ..= 36 => (letter - 11 + b'A') as char,
      37 ..= 62 => (letter - 37 + b'a') as char,
      63 => '_',
      _ => panic!("Invalid letter in name"),
    });
  }
  name
}

pub fn u32_to_letters(num: u32) -> Vec<u8> {
  let mut letters = Vec::new();
  let mut num = num;
  while num > 0 {
    letters.push((num % 64) as u8);
    num /= 64;
  }
  letters.reverse();
  letters
}

pub fn letters_to_u32(letters: Vec<u8>) -> u32 {
  let mut num = 0;
  for letter in letters {
    num = num * 64 + letter as u32;
  }
  num
}

pub fn name_to_u32(name: &str) -> u32 {
  letters_to_u32(name_to_letters(name))
}

pub fn u32_to_name(num: u32) -> String {
  letters_to_name(u32_to_letters(num))
}

impl fmt::Display for LNet {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    writeln!(f, "  $ {}", self.root)?;
    for (a, b) in &self.acts {
      writeln!(f, "  & {}", a)?;
      writeln!(f, "  ~ {}", b)?;
    }
    Ok(())
  }
}

impl fmt::Display for LTree {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match self {
      LTree::Era => write!(f, "*"),
      LTree::Nod { tag, lft, rgt } => write!(f, "({} {} {})", tag - CON, lft, rgt),
      LTree::Var { nam } => write!(f, "{nam}"),
      LTree::Num { val } => write!(f, "{val}"),
      LTree::Ref { nam } => write!(f, "@{}", letters_to_name(u32_to_letters(*nam))),
    }
  }
}

impl fmt::Display for Book {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    for (id, net) in &self.defs {
      writeln!(f, "{} =", u32_to_name(**id))?;
      writeln!(f, "{}\n", net)?;
    }
    Ok(())
  }
}

pub type Tag = u8;
pub type Val = u32;

pub const NIL: Tag = 0x0; // empty node
pub const REF: Tag = 0x1; // reference to a definition (closed net)
pub const NUM: Tag = 0x2; // unboxed number
pub const ERA: Tag = 0x3; // unboxed eraser
pub const VRR: Tag = 0x4; // variable pointing to root
pub const VR1: Tag = 0x5; // variable pointing to aux1 port of node
pub const VR2: Tag = 0x6; // variable pointing to aux2 port of node
pub const RDT: Tag = 0x7; // redirection to root
pub const RD1: Tag = 0x8; // redirection to aux1 port of node
pub const RD2: Tag = 0x9; // redirection to aux2 port of node
pub const CON: Tag = 0xA; // points to main port of con node
pub const DUP: Tag = 0xB; // points to main port of dup node; higher labels also dups
