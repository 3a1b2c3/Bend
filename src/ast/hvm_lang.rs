use super::{DefId, Name};
use bimap::{BiHashMap, Overwritten};
use itertools::Itertools;
use std::fmt;

#[cfg(feature = "nums")]
use hvm_core::{opx_to_string, OP};

#[derive(Debug, Clone, Default)]
pub struct DefNames {
  map: BiHashMap<DefId, Name>,
  id_count: DefId,
}

#[derive(Debug, Clone, Default)]
pub struct DefinitionBook {
  pub def_names: DefNames,
  pub defs: Vec<Definition>,
}

#[derive(Debug, Clone)]
pub struct Definition {
  pub def_id: DefId,
  pub rules: Vec<Rule>,
}

#[derive(Debug, Clone)]
pub struct Rule {
  pub def_id: DefId,
  pub pats: Vec<Pattern>,
  pub body: Term,
}

#[derive(Debug, Clone)]
pub enum Pattern {
  Ctr(Name, Vec<Pattern>),
  Var(Option<Name>),
  #[cfg(feature = "nums")]
  U32(u32),
  #[cfg(feature = "nums")]
  I32(i32),
}

#[derive(Debug, Clone)]
pub enum Term {
  Lam {
    nam: Option<Name>,
    bod: Box<Term>,
  },
  Var {
    nam: Name,
  },
  /// Like a scopeless lambda, where the variable can occur outside the body
  Chn {
    nam: Name,
    bod: Box<Term>,
  },
  /// The use of a Channel variable.
  Lnk {
    nam: Name,
  },
  Let {
    nam: Name,
    val: Box<Term>,
    nxt: Box<Term>,
  },
  Ref {
    def_id: DefId,
  },
  App {
    fun: Box<Term>,
    arg: Box<Term>,
  },
  Dup {
    fst: Option<Name>,
    snd: Option<Name>,
    val: Box<Term>,
    nxt: Box<Term>,
  },
  Sup {
    fst: Box<Term>,
    snd: Box<Term>,
  },
  Era,
  #[cfg(feature = "nums")]
  U32 {
    val: u32,
  },
  #[cfg(feature = "nums")]
  I32 {
    val: i32,
  },
  #[cfg(feature = "nums")]
  /// A numeric operation between built-in numbers.
  Opx {
    op: OP,
    fst: Box<Term>,
    snd: Box<Term>,
  },
}

impl DefinitionBook {
  pub fn new() -> Self {
    Default::default()
  }
}

impl DefNames {
  pub fn new() -> Self {
    Default::default()
  }

  pub fn name(&self, def_id: &DefId) -> Option<&Name> {
    self.map.get_by_left(def_id)
  }

  pub fn def_id(&self, name: &Name) -> Option<DefId> {
    self.map.get_by_right(name).copied()
  }

  pub fn contains_name(&self, name: &Name) -> bool {
    self.map.contains_right(name)
  }

  pub fn contains_def_id(&self, def_id: &DefId) -> bool {
    self.map.contains_left(def_id)
  }

  pub fn insert(&mut self, name: Name) -> DefId {
    let def_id = self.id_count;
    self.id_count.0 += 1;
    match self.map.insert(def_id, name) {
      Overwritten::Neither => def_id,
      _ => todo!("Overwritting name-id pairs not supported"),
    }
  }
}

impl Term {
  pub fn to_string(&self, def_names: &DefNames) -> String {
    match self {
      Term::Lam { nam, bod } => {
        format!("λ{} {}", nam.clone().unwrap_or(Name::new("*")), bod.to_string(def_names))
      }
      Term::Var { nam } => format!("{nam}"),
      Term::Chn { nam, bod } => format!("λ${} {}", nam, bod.to_string(def_names)),
      Term::Lnk { nam } => format!("${nam}"),
      Term::Let { nam, val, nxt } => {
        format!("let {} = {}; {}", nam, val.to_string(def_names), nxt.to_string(def_names))
      }
      Term::Ref { def_id } => format!("{}", def_names.name(def_id).unwrap()),
      Term::App { fun, arg } => format!("({} {})", fun.to_string(def_names), arg.to_string(def_names)),
      Term::Dup { fst, snd, val, nxt } => format!(
        "dup {} {} = {}; {}",
        fst.as_ref().map(|x| x.as_str()).unwrap_or("*"),
        snd.as_ref().map(|x| x.as_str()).unwrap_or("*"),
        val.to_string(def_names),
        nxt.to_string(def_names)
      ),
      Term::Sup { fst, snd } => format!("{{{} {}}}", fst.to_string(def_names), snd.to_string(def_names)),
      Term::Era => "*".to_string(),
      #[cfg(feature = "nums")]
      Term::U32 { val } => format!("{val}"),
      #[cfg(feature = "nums")]
      Term::I32 { val } => format!("{val:+}"),
      #[cfg(feature = "nums")]
      Term::Opx { op, fst, snd } => {
        format!("({} {} {})", opx_to_string(op), fst.to_string(def_names), snd.to_string(def_names))
      }
    }
  }

  /// Make a call term by folding args around a called function term with applications.
  pub fn call(called: Term, args: impl IntoIterator<Item = Term>) -> Self {
    args.into_iter().fold(called, |acc, arg| Term::App { fun: Box::new(acc), arg: Box::new(arg) })
  }

  /// Substitute the occurences of a variable in a term with the given term.
  pub fn subst(&mut self, from: &Name, to: &Term) {
    match self {
      Term::Lam { nam: None, bod } => bod.subst(from, to),
      Term::Lam { nam: Some(nam), bod } if nam == from => (),
      Term::Lam { nam: Some(nam), bod } => bod.subst(from, to),
      Term::Var { nam } if nam == from => *self = to.clone(),
      Term::Var { nam } => (),
      // Only substitute scoped variables.
      Term::Chn { nam, bod } => bod.subst(from, to),
      Term::Lnk { nam } => (),
      Term::Let { nam, val, nxt } => {
        val.subst(from, to);
        if nam != from {
          nxt.subst(from, to);
        }
      }
      Term::Ref { def_id } => (),
      Term::App { fun, arg } => {
        fun.subst(from, to);
        arg.subst(from, to);
      }
      Term::Dup { fst, snd, val, nxt } => {
        val.subst(from, to);
        if fst.as_ref().map_or(true, |fst| fst == from) && fst.as_ref().map_or(true, |snd| snd == from) {
          nxt.subst(from, to);
        }
      }
      Term::Sup { fst, snd } => {
        fst.subst(from, to);
        snd.subst(from, to);
      }
      Term::Era => (),
    }
  }
}

impl Rule {
  pub fn to_string(&self, def_names: &DefNames) -> String {
    let Rule { def_id, pats, body } = self;
    format!(
      "({}{}) = {}",
      def_names.name(def_id).unwrap(),
      pats.iter().map(|x| format!(" {x}")).join(""),
      body.to_string(def_names)
    )
  }
}

impl Definition {
  pub fn to_string(&self, def_names: &DefNames) -> String {
    self.rules.iter().map(|x| x.to_string(def_names)).join("\n")
  }
}

impl From<&Pattern> for Term {
  fn from(value: &Pattern) -> Self {
    match value {
      Pattern::Ctr(nam, args) => Term::call(Term::Var { nam: nam.clone() }, args.iter().map(Term::from)),
      Pattern::Var(nam) => Term::Var { nam: Name::new(nam.as_ref().map(|x| x.as_str()).unwrap_or("_")) },
      #[cfg(feature = "nums")]
      Pattern::U32(num) => Term::U32 { num },
      #[cfg(feature = "nums")]
      Pattern::I32(num) => Term::I32 { num },
    }
  }
}

impl fmt::Display for DefinitionBook {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
      write!(f, "{}", self.defs.iter().map(|x| x.to_string(&self.def_names)).join("\n\n"))
    }
}

impl fmt::Display for Pattern {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    match self {
      Pattern::Ctr(name, pats) => write!(f, "({}{})", name, pats.iter().map(|p| format!(" {p}")).join("")),
      Pattern::Var(nam) => write!(f, "{}", nam.as_ref().map(|x| x.as_str()).unwrap_or("*")),
      #[cfg(feature = "nums")]
      Pattern::U32(num) => write!(f, "{num}"),
      #[cfg(feature = "nums")]
      Pattern::I32(num) => write!(f, "{num:+}"),
    }
  }
}
