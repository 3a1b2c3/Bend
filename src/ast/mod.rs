use std::collections::HashMap;

use shrinkwraprs::Shrinkwrap;

#[derive(Debug, Clone, Default)]
pub struct DefinitionBook {
  pub definitions: HashMap<Name, Definition>,
}

#[derive(Debug, Clone)]
pub struct Definition {
  pub name: Name,
  pub rules: Vec<Rule>,
}

#[derive(Debug, Clone)]
pub struct Rule {
  pub name: Name,
  pub pats: Vec<Pattern>,
  pub body: Term,
}

#[derive(Debug, Clone)]
pub enum Pattern {
  Ctr(Name, Vec<Pattern>),
  Num(Number),
  Var(Name),
}

#[derive(Debug, Clone)]
pub enum Term {
  Lam { name: Name, body: Box<Term> },
  Var { name: Name },
  App { func: Box<Term>, arg: Box<Term> },
  Dup { fst: Name, snd: Name, val: Box<Term>, next: Box<Term> },
  Def { name: Name },
  Num { val: Number },
  NumOp { op: NumOper, fst: Box<Term>, snd: Box<Term> },
}

#[derive(Debug, PartialEq, Eq, Clone, Shrinkwrap, Hash)]
pub struct Name(pub String);

// TODO: Accept other number types
#[derive(Debug, PartialEq, Eq, Clone, Copy, Shrinkwrap)]
pub struct Number(pub u16);

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum NumOper {
  Add,
  Sub,
  Mul,
  Div,
  Mod,
  And,
  Or,
  Xor,
  Shl,
  Shr,
  Ltn,
  Lte,
  Gtn,
  Gte,
  Eql,
  Neq,
}

impl DefinitionBook {
  pub fn new() -> Self {
    Default::default()
  }
}
