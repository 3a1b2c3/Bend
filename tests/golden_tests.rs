use hvmc::ast::{parse_net, show_net};
use hvml::{
  compile_book, encode_pattern_matching,
  net::{hvmc_to_net::hvmc_to_net, net_to_hvmc::net_to_hvmc},
  run_book,
  term::{
    load_book::do_parse_book, net_to_term::net_to_term, parser::parse_term, term_to_compat_net, Book, DefId,
    Term,
  },
  Opts, WarningOpts,
};
use insta::assert_snapshot;
use itertools::Itertools;
use std::{
  collections::HashMap,
  fs,
  path::{Path, PathBuf},
  sync::{Arc, RwLock},
};
use stdext::function_name;
use walkdir::WalkDir;

fn do_parse_term(code: &str) -> Result<Term, String> {
  parse_term(code).map_err(|errs| errs.into_iter().map(|e| e.to_string()).join("\n"))
}

fn do_parse_net(code: &str) -> Result<hvmc::ast::Net, String> {
  parse_net(&mut code.chars().peekable())
}

const TESTS_PATH: &str = "/tests/golden_tests/";

fn run_single_golden_test(
  path: &Path,
  run: &dyn Fn(&str, &Path) -> Result<String, String>,
) -> Result<(), String> {
  let code = fs::read_to_string(path).map_err(|e| e.to_string())?;
  let file_name = path.to_str().and_then(|path| path.rsplit_once(TESTS_PATH)).unwrap().1;

  // unfortunately we need to do this
  let file_path = format!("{}{}", &TESTS_PATH[1 ..], file_name);
  let file_path = Path::new(&file_path);

  let result: String = run(&code, file_path).unwrap_or_else(|err| err);

  let mut settings = insta::Settings::clone_current();
  settings.set_prepend_module_to_snapshot(false);
  settings.set_omit_expression(true);
  settings.set_input_file(path);

  settings.bind(|| {
    assert_snapshot!(file_name, result);
  });

  Ok(())
}

fn run_golden_test_dir(test_name: &str, run: &dyn Fn(&str, &Path) -> Result<String, String>) {
  let root = PathBuf::from(format!(
    "{}{TESTS_PATH}{}",
    env!("CARGO_MANIFEST_DIR"),
    test_name.rsplit_once(':').unwrap().1
  ));

  let walker = WalkDir::new(&root).sort_by_file_name().max_depth(2).into_iter().filter_entry(|e| {
    let path = e.path();
    path == root || (path.is_file() && path.extension().is_some_and(|x| x == "hvm"))
  });

  for entry in walker {
    let entry = entry.unwrap();
    let path = entry.path();
    if path.is_file() {
      eprintln!("Testing {}", path.display());
      run_single_golden_test(path, run).unwrap();
    }
  }
}

#[test]
fn compile_term() {
  run_golden_test_dir(function_name!(), &|code, _| {
    let mut term = do_parse_term(code)?;
    term.check_unbound_vars(&mut HashMap::new())?;
    term.make_var_names_unique();
    term.linearize_vars();
    let compat_net = term_to_compat_net(&term, &mut Default::default());
    let net = net_to_hvmc(&compat_net, &|def_id| def_id.to_internal())?;

    let result = show_net(&net);

    Ok(result)
  })
}

#[test]
fn compile_file_o_all() {
  run_golden_test_dir(function_name!(), &|code, path| {
    let mut book = do_parse_book(code, path)?;
    let compiled = compile_book(&mut book, Opts::heavy())?;
    Ok(format!("{:?}", compiled))
  })
}
#[test]
fn compile_file() {
  run_golden_test_dir(function_name!(), &|code, path| {
    let mut book = do_parse_book(code, path)?;
    let compiled = compile_book(&mut book, Opts::light())?;
    Ok(format!("{:?}", compiled))
  })
}

#[test]
fn run_file() {
  run_golden_test_dir(function_name!(), &|code, path| {
    let book = do_parse_book(code, path)?;
    // 1 million nodes for the test runtime. Smaller doesn't seem to make it any faster
    let (res, def_names, info) =
      run_book(book, 1 << 20, true, false, false, WarningOpts::deny_all(), Opts::heavy())?;
    Ok(format!("{}{}", info.readback_errors.display(&def_names), res.display(&def_names)))
  })
}

#[test]
fn readback_lnet() {
  run_golden_test_dir(function_name!(), &|code, _| {
    let net = do_parse_net(code)?;
    let book = Book::default();
    let compat_net = hvmc_to_net(&net, &DefId::from_internal);
    let (term, errors) = net_to_term(&compat_net, &book, &Default::default(), false);
    Ok(format!("{}{}", errors.display(&book.def_names), term.display(&book.def_names)))
  })
}

#[test]
fn flatten_rules() {
  run_golden_test_dir(function_name!(), &|code, path| {
    let mut book = do_parse_book(code, path)?;
    let main = book.check_has_main().ok();
    book.check_shared_names()?;
    book.encode_builtins();
    book.resolve_ctrs_in_pats();
    book.generate_scott_adts();
    book.desugar_let_destructors();
    book.desugar_implicit_match_binds();
    book.check_unbound_pats()?;
    book.extract_adt_matches(&mut Vec::new())?;
    book.flatten_rules();
    book.prune(main, false, &mut Vec::new());
    Ok(book.to_string())
  })
}

#[test]
fn parse_file() {
  run_golden_test_dir(function_name!(), &|code, path| {
    let book = do_parse_book(code, path)?;
    Ok(book.to_string())
  })
}

#[test]
fn encode_pattern_match() {
  run_golden_test_dir(function_name!(), &|code, path| {
    let mut book = do_parse_book(code, path)?;
    let main = book.check_has_main().ok();
    book.check_shared_names()?;
    book.encode_builtins();
    book.generate_scott_adts();
    book.resolve_refs()?;
    encode_pattern_matching(&mut book, &mut Vec::new())?;
    book.prune(main, false, &mut Vec::new());
    Ok(book.to_string())
  })
}

#[test]
#[ignore = "To not delay golden tests execution"]
fn hangs() {
  let expected_normalization_time = 1;

  run_golden_test_dir(function_name!(), &|code, path| {
    let book = do_parse_book(code, path)?;

    let lck = Arc::new(RwLock::new(false));
    let got = lck.clone();
    std::thread::spawn(move || {
      let _ = run_book(book, 1 << 20, true, false, false, WarningOpts::deny_all(), Opts::heavy());
      *got.write().unwrap() = true;
    });
    std::thread::sleep(std::time::Duration::from_secs(expected_normalization_time));

    if !*lck.read().unwrap() { Ok("Hangs".into()) } else { Err("Doesn't hang".into()) }
  })
}
