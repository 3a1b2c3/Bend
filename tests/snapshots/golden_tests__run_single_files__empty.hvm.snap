---
source: tests/golden_tests.rs
expression: result
---
File has no 'main' definition
